### PR TITLE
Base implementation of autocomplete #849

### PR DESCRIPTION
**Pull request by ricardojlrufino**
Start with the minimum base for autocomplete support.
The functionality is initialized with the key: Ctrl + Space
A test class has been created: FakeCompletionProvider, to enable usability testing on other platforms

Implementation can be added through the class extension:
Cc.arduino.autocomplete.CompletionProvider

### Tests:
[X] Ubuntu 16.04 64

### TODO:
[ ] Discovery of implementations (using preferences or using Java SPI ) ?

### CODE REVIEW

**Comment by facchinm** [comment]
@ArduinoBot build this please

**Comment by tigoe** [comment]
I find this confusing, and couldn't get it to autocomplete. When i press control-space (an unusual key combo for OSX users), I get a popup box that interrupts my typing, and when I dismiss it with the enter key, it puts the line number in my code. 

What I assumed when I saw "autocomplete" was something like what Atom (https://atom.io/) or Coda (https://panic.com/coda/) do: when the editor knows the context from the file type (e.g. .html, .js) it auto-completes from keywords when you tab or hit enter, there's no additional popup box, but there is a very small hovering text tip that doesn't block your view, which you can turn on or off.. It's not invasive, you can ignore it if you choose, or you can complete it with one keystroke. To me, that is what I would expect in an auto-complete. I don't see how that relates to what this is doing. Can you explain further what you're intent is?

**Comment by cmaglie** [comment]
@tigoe 
this PR is aimed to do the **only** the GUI part first, so the popup is not populated with actual code-completions but with some placeholders. @ricardojlrufino made this after my request here #849, because his first submission of a full implementation of autocomplete (here https://github.com/arduino/Arduino/pull/3484) is way too complex to handle all at once and touch many aspects of the IDE that are unrelated with the autocompletion.

So this is a try to go with an incremental approach, once done the GUI part we can focus on the code-completion engine that will be the tricky part to get right.

**Comment by tigoe** [comment]
I see, thank you for the explanation. In which case, my thoughts on the GUI are as follows: the popup window is too big and invasive. I would do something smaller, that takes less screen real estate. And the key combination is uncomfortable for OSX users. If it requires a command-key combo, I would use control as the modifier on windows and Linux, but command as the modifier on OSX, as is standard. Though ideally, I would not require a modifier except to turn the mode on or off; I'd use enter or tab. like other auto-corrects fo.

**Comment by ricardojlrufino** [comment]
@cmaglie  My idea was to follow their guidelines, to have a cleaner interface.
But in fact, I think it's best to use what we already have ...

> Is this the autocomplete library here?
> https://github.com/bobbylight/AutoComplete

YES

@tigoe In that first moment, I find it interesting to focus on the functional aspects. The visual part and usability can have its specific moment.
These changes could be made in the bobbylight / AutoComplete project, I believe. I could try doing them (although I'm not an OSX user .. =])

**Comment by tigoe** [comment]
OK. When you're dealing with the visual/usability, I'll pay attention again, because that is the most important aspect to me.

**Comment by ricardojlrufino** [comment]
Now, basically, there have been two classes ...

SketchCompletionProvider, is the basis for providing autocomplete services for various contexts: Code, Strings, documentation.

Specifically the 'code' that is what interests us at that moment, will be offered by classses that extend: BaseCCompletionProvider

Currently the two classes do not have much, but they will be necessary as we implement new functionalities.

For reference, look at my previous code:

**Comment by ricardojlrufino** [comment]
Hi, @cmaglie. Did you have time to do the review?

File: app/src/cc/arduino/autocomplete/BaseCCompletionProvider.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package cc.arduino.autocomplete;
+
+import org.fife.ui.autocomplete.DefaultCompletionProvider;
+
+/**
+ * Base completion provider for C/C++.
+ * @author Ricardo JL Rufino (ricardo@criativasoft.com.br)
```
Comment (cmaglie) [inline app/src/cc/arduino/autocomplete/BaseCCompletionProvider.java]: May you add a GNU GPL license on top of all the files and move the credits there?
You can copy the header from any of the other files.

File: app/src/processing/app/EditorTab.java [hunk 2]
```diff
@@ -106,6 +107,8 @@ public EditorTab(Editor editor, SketchFile file, String contents)
     file.setStorage(this);
     applyPreferences();
     add(scrollPane, BorderLayout.CENTER);
+    
+    textarea.setupAutoComplete(editor.getSketch(), new FakeCompletionProvider());
```
Comment (cmaglie) [inline app/src/processing/app/EditorTab.java]: Ok some thoughts about this:

- this call could be commented out, so we can merge this PR without affecting the IDE behavior for now.

- what about adding a method like `Editor.setAutocompletionProvider(CompletionProvider)`? this way we can implement the completion-engine into a tool...

- moreover I think that this call can be inlined (and the method `textarea.setupAutoComplete` removed).

**Comment by ricardojlrufino** [comment]
Hi @cmaglie , I followed your suggestions, what's the next step?

**Comment by ricardojlrufino** [comment]
@cmaglie PING ! =)

**Comment by facchinm** [comment]
The relevant bits of this PR has been merged in beta-1.9 branch, closing as merged

**Pull request closed without merging**